File: dynamic_listings/python/default/containers/fsm_usage.py
class Enemy:
    """
    Represents a simple enemy
    """
    # ...
    PURSUETIME: float = 10.0
    position_x: float = 0.0
    position_y: float = 0.0
    pursue_timer: Timer = Timer()
    brain: FSM = FSM()

    def __init__(self, x: float, y: float):
        """
        Constructor
        """
        self.position_x = x
        self.position_y = y
        self.brain.setState(self.patrol)

    def sees(self, other: Entity) -> bool:
        """
        Implements logic for the "sight" of the enemy
        """
        # ...
        pass

    def patrol(self, dt: float) -> None:
        # Normal patrolling of the enemy
        # Move, turn, path find...
        if self.sees(player):
            # ...
            # Pursue for xx seconds
            self.pursue_timer.set(self.PURSUETIME)
            self.pursue_timer.start()
            # Change FSM State
            self.brain.setState(self.pursue)

    def pursue(self, dt: float) -> None:
        # Tries to pursue the enemy
        if self.sees(player):
            # Continue Pursuing, by resetting the timer
            self.pursue_timer.set(self.PURSUETIME)
            # ...
        # ...
        # If the enemy is not in sight for xx seconds
        if (self.pursue_timer.is_finished()):
            # go back to patrolling
            self.brain.setState(self.patrol)

    def update(self, dt: float) -> None:
        # The enemy update function
        # ...
        self.pursue_timer.update(dt)
        self.brain.update(dt)
        # ...

File: dynamic_listings/C++/default/containers/fsm_usage.cpp
class Enemy{
    /*
     * Represents a simple enemy
     */
    // ...
    const float PURSUETIME = 10.0;
    float position_x = 0.0;
    float position_y = 0.0;
    Timer pursue_timer = Timer();
    FSM brain = FSM();

    Enemy(float x, float y){
        /*
         * Constructor
         */
        position_x = x;
        position_y = y;
        brain.setState(patrol);
    }

    bool sees(Entity other){
        /*
         *Implements logic for the "sight" of the enemy
         */
        // ...
    }

    void patrol(float dt){
        // Normal patrolling of the enemy
        // Move, turn, path find...
        if (sees(player)){
            // ...
            // Pursue for xx seconds
            pursue_timer.set(PURSUETIME);
            pursue_timer.start();
            // Change FSM State
            brain.setState(pursue);
        }
    }

    void pursue(float dt){
        // Tries to pursue the enemy
        if (sees(player)){
            // Continue Pursuing, by resetting the timer
            pursue_timer.set(PURSUETIME);
            // ...
        }
        // ...
        // If the enemy is not in sight for xx seconds
        if (pursue_timer.is_finished()){
            // go back to patrolling
            brain.setState(patrol);
        }
    }

    void update(float dt){
        // The enemy update function
        // ...
        pursue_timer.update(dt)
        brain.update(dt)
        // ...
    }
};
